File: byteback-dummy/scala/src/main/scala/byteback/dummy/complete/LinearSearch.scala
package byteback.dummy.complete;

import byteback.annotations.Contract._;
import byteback.annotations.Operator._;
import byteback.annotations.Operator.{eq => equal};

class LinearSearch {

  @Condition
  def bounded_indices(a: Array[Int], n: Int, left: Int, right: Int): Boolean = {
    return lte(0, left) & lte(left, right) & lte(right, a.length);
  }

  @Condition
  def result_is_index(a: Array[Int], n: Int, left: Int, right: Int, returns: Int): Boolean = {
    return implies(lte(0, returns), equal(a(returns), n));
  }

  @Require("bounded_indices")
  @Ensure("result_is_index")
  def apply(a: Array[Int], n: Int, left: Int, right: Int): Int = {
    var i: Int = left;

    while (i < right) {
			invariant(lte(left, i) & lte(i, right));

      if (a(i) == n) {
        return i;
      }

      i = i + 1;
    }

    return -1;
  }

}

File: byteback-dummy/java8/src/main/java/byteback/dummy/complete/LinearSearch.java
package byteback.dummy.complete;

import static byteback.annotations.Contract.*;
import static byteback.annotations.Operator.*;

public class LinearSearch {

	@Condition
	public static boolean bounded_indices(int a[], int n, int left, int right) {
		return lte(0, left) & lte(left, right) & lte(right, a.length);
	}

	@Condition
	public static boolean result_is_index(int a[], int n, int left, int right, int returns) {
		return implies(lte(0, returns), eq(a[returns], n));
	}

	@Require("bounded_indices")
	@Ensure("result_is_index")
	public static int search(int a[], int n, int left, int right) {

		for (int i = left; i < right; ++i) {
			invariant(lte(left, i) & lte(i, right));

			if (a[i] == n) {
				return i;
			}
		}

		return -1;
	}
}
